lib/ClangImporter/ImportName.cpp

```diff
@@ -1741,11 +1741,10 @@ ImportedName NameImporter::importNameImpl(const clang::NamedDecl *D,
       break;
     case clang::OverloadedOperatorKind::OO_Subscript: {
       auto returnType = functionDecl->getReturnType();
-      if (!returnType->isReferenceType()) {
-        // TODO: support non-reference return types (SR-14351)
-        return ImportedName();
-      }
-      if (returnType->getPointeeType().isConstQualified()) {
+      if (!returnType->isReferenceType() ||
+          returnType->getPointeeType().isConstQualified()) {
```

Comment:
This condition looks a bit suspicious – what if `operator[]` returns a pointer instead of a reference?

For example, this code
```cpp
struct PointerStuff {
private:
  int *values[3];

public:
  const int* operator[](int x) const {
    return values[x];
  }

  int* operator[](int x) {
    return values[x];
  }
};
```
gets imported as
```swift
public struct PointerStuff {
  public init()
  @available(*, unavailable, message: "use subscript")
  public mutating func __operatorSubscriptConst(_ x: Int32) -> UnsafePointer<Int32>!
  @available(*, unavailable, message: "use subscript")
  public mutating func __operatorSubscriptConst(_ x: Int32) -> UnsafeMutablePointer<Int32>!
  public subscript(x: Int32) -> UnsafeMutablePointer<Int32>? { mutating get }
}
```

Notice that `__operatorSubscriptConst` is defined twice, and the subscript has a weird return type (which BTW depends on the order in which the operators are declared in C++).

Perhaps it would be nicer to avoid importing such operators altogether?

Reply:
@egorzhdan Nice observation! I have to think about this one a bit.

Reply:
@egorzhdan I decided to allow `T *operator[]` to gave a setter depending on constness. One problem I am still hitting is the Optional to Optional assignment (hits an assert). So while `arr[0]!.a = 64` might work, arr[0] = arr[0] asserts for the pointer case.

Reply:
@plotfi  this approach looks good to me! With the optionals I think you might need to do either 
1. adjust the logic that imports `__operatorSubscriptConst` so that its return type is `UnsafePointer<Int32>?` instead of `UnsafePointer<Int32>!` or
2. add some kind of conversion between optional & an implicitly-unwrapped optional in the `get`/`set` accessor synthesis logic (`synthesizeSubscriptGetterBody`/`synthesizeSubscriptSetterBody`).

Reply:
@egorzhdan Ah yeah, me and noticed that the optional types were not lining up. ie, the following: `int a = 64; struct S { int *operator[](int x) { return &a; }  };` was producing:

```
struct S {
  init()
  @available(*, unavailable, message: "use subscript")
  mutating func __operatorSubscript(_ x: Int32) -> UnsafeMutablePointer<Int32>!
  subscript(x: Int32) -> UnsafeMutablePointer<Int32>? { mutating get set }
}
``` 

I tried changing the subscript type to a `UnsafeMutablePointer<Int32>!` but that did not work. So then you mean, the `__operatorSubscript` function return type should be a  `UnsafeMutablePointer<Int32?` then?

Reply:
> I tried changing the subscript type to a UnsafeMutablePointer<Int32>! but that did not work.

Hmm, that in theory should work fine... and it feels like a better idea than what I had in mind (returning `UnsafeMutablePointer<Int32>?`). Perhaps the issue is with the return type unwrapping in `ImportDecl.cpp:7666`? With this approach we'll probably need to distinguish between pointers & references there (we currently don't), since we'd like to unwrap references (`UnsafeMutablePointer<T>` -> `T`) but leave pointers as is.

Reply:
@egorzhdan Ah yes, thank you. I managed to kinda hard code some unwrapping at around `ImportDecl.cpp:7666`. It seems to kind of work, but still have some issues generating IR later due to the a type mismatch between the subscript's `UnsafeMutablePointer<Int32>` and `__operatorSubscript(_ x: Int32) -> UnsafeMutablePointer<Int32>!`. Trying to figure out how to unwrap the `__operatorSubscript` now

Reply:
(Sorry for not responding sooner...)

@plotfi I think that's actually the wrong way to go about it. I think we should be trying to discover why the subscript's return type isn't being implicitly unwrapped.

So, both return types here are optional (which makes sense, because they both have either a `?` or `!` after them). But, only the member (`__operatorSubscript`) is being implicitly unwrapped (hence the `!` and not the `?`). The member's return type is implicitly unwrapped because of [this line in importFunctionDecl](https://github.com/apple/swift/blob/main/lib/ClangImporter/ImportDecl.cpp#L4191-L4192). 

[Note that we do the same thing for synthesized constructors.](https://github.com/apple/swift/blob/main/lib/ClangImporter/ImportDecl.cpp#L1420)

But, we don't do it for synthesized subscripts. We need to do it for them as well. Simply adding:
```
Impl.recordImplicitUnwrapForDecl(subscript,
                                   getterImpl->isImplicitlyUnwrappedOptional());
```
to the end of `makeSubscript` will make this work properly, so we get:
```
struct MystructPtrByVal
  mutating func __operatorSubscript(_ x: Int32) -> UnsafeMutablePointer<Int32>!
  subscript(x: Int32) -> (Int32) -> UnsafeMutablePointer<Int32>! { mutating get set }
```

Reply:
@zoecarver Ah excellent, That did the trick! I was playing around with some similar things in makeSubscript but I think I had put them too early in the function. Also, I think we discussed this, but I got to some point in my debugging of this case where I was seeing the ide tool print `UnsafeMutablePointer<Int32>!` for both the __operatorSubscript and the subscript function but still hit the assert. Is there something odd going on with the dumper?